### PR TITLE
Scheduler: remove pkg/apis/core/field_constants.go

### PR DESCRIPTION
**Pull request by xiaoanyunfei**
**What type of PR is this?**

 /kind cleanup

**What this PR does / why we need it**:
Scheduler: remove pkg/apis/core/field_constants.go

**Which issue(s) this PR fixes**:

Ref https://github.com/kubernetes/kubernetes/issues/89930

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig scheduling

### CODE REVIEW

**Review by damemi — APPROVED** [review]
This looks like a very thorough cleanup to me, no reason not to have the default scheduler name in the external api imo.
/lgtm

File: staging/src/k8s.io/api/core/v1/types.go
Comment (neolit123) [inline staging/src/k8s.io/api/core/v1/types.go:5916]: needs an export comment if added here.
instead of moving one of the object fields constants to here it makes sense to move all of them to fragment less.

yet, unclear whether this is the right location.
given "metadata.*" as a field selector path is unlikely to change, these field constants field a bit redundant and makes more sense to me to define them in locations that need them.

Comment (xiaoanyunfei) [reply staging/src/k8s.io/api/core/v1/types.go]: done

**Comment by tanjunchen** [comment]
Migrated well
/lgtm

**Comment by alculquicondor** [comment]
/approve
for pkg/scheduler

**Comment by damemi** [comment]
Bump, @smarterclayton do you have a chance to approve this? It's just moving usage of the default scheduler name to the external api, but it touches all the areas that were referencing it so the list of required owners is huge

**Review by damemi — COMMENTED** [review]
@xiaoanyunfei sorry this got lost, are you still open to rebasing and picking it back up?

I'm not sure why review stalled, I don't see any early objections to it. Perhaps just the wide list of reviewers, but it's a pretty straightforward move of some constants. Would like to see this get going again :)

**Comment by xiaoanyunfei** [comment]
@damemi  I have rebased the code

**Review by damemi — COMMENTED** [review]
Thanks @xiaoanyunfei, this still looks good to me. Unfortunately we are past code freeze so it will be 2 weeks before this can actually merge but that doesn't mean we can't get some more review on this from the appropriate approvers.

To me it seems like the main point of this PR is actually to move `pkg/apis/core/field_constants.go` to the external API, which is apparently noted as a todo in one of the changes you made. It might be appropriate to retitle this PR to something that reflects that.

Since the list of owners who will be affected by this move is so large, it might make sense to first get a general +1 from some API folks (@liggitt or @smarterclayton) to ensure this is a feasible move. That file has a comment indicating that the constants within it are "specific to the internal representation" but I'm interested if there's actually anything really specific. It also hasn't been modified in 3 years, so this may be reasonable cleanup.

If they agree with moving `field_constants.go`, then we can seek out approvers from all the areas affected by this (though I don't think many will object to the changes that affect them).

File: staging/src/k8s.io/api/core/v1/field_constants.go
Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/field_constants.go:20]: Taking the fields one by one:

```
	NodeUnschedulableField = "spec.unschedulable"
```
this is related to v1 Nodes, but is unreferenced. I would drop it rather than export it.

```
	ObjectNameField        = "metadata.name"
```
this is not only external but is cross-version... I could see it moving to `staging/src/k8s.io/apimachinery/pkg/apis/meta/v1`

```
	PodHostField           = "spec.nodeName"
	PodStatusField         = "status.phase"
```
these are related to v1 Pods, but we don't typically export field selector constants... what would it look like to use inline strings instead of exporting as an API constant?

```
	SecretTypeField        = "type"
```
this is related to v1 Secrets, but we don't typically export field selector constants... what would it look like to use inline strings instead of exporting as an API constant?

```
	EventReasonField                  = "action"
	EventSourceField                  = "reportingComponent"
	EventTypeField                    = "type"
	EventInvolvedKindField            = "involvedObject.kind"
	EventInvolvedNamespaceField       = "involvedObject.namespace"
	EventInvolvedNameField            = "involvedObject.name"
	EventInvolvedUIDField             = "involvedObject.uid"
	EventInvolvedAPIVersionField      = "involvedObject.apiVersion"
	EventInvolvedResourceVersionField = "involvedObject.resourceVersion"
	EventInvolvedFieldPathField       = "involvedObject.fieldPath"
```
These are not referenced anywhere, but match values used in core/v1 event field selectors https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/apis/core/v1/conversion.go#L426-L448) and internally (https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/registry/core/event/strategy.go#L111-L123)

The events.k8s.io API converts field names to these as well: 
* https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/apis/events/v1/conversion.go#L63-L77
* https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/apis/events/v1beta1/conversion.go#L63-L77

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/field_constants.go:20]: We don't typically export constants for field selector keys. `ObjectNameField` is the only constant here I would expect us to export, and it seems like it should be in the meta/v1 package more than the core/v1 package.

File: pkg/apis/core/types.go
Comment (xiaoanyunfei) [reply pkg/apis/core/types.go:5336]: We already have `k8s.io/api/core/v1.DefaultSchedulerName`, and I just want to prevent this being referenced by other packages

Comment (xiaoanyunfei) [reply pkg/apis/core/types.go:5336]: > why drop the constant from this package?

We already have k8s.io/api/core/v1.DefaultSchedulerName, and I just want to prevent this being referenced by other packages

File: staging/src/k8s.io/api/core/v1/field_constants.go
Comment (xiaoanyunfei) [reply staging/src/k8s.io/api/core/v1/field_constants.go]: >  ObjectNameField        = "metadata.name"

I moved `ObjectNameField` to `staging/src/k8s.io/apimachinery/pkg/apis/meta/v1`


> PodHostField    = "spec.nodeName"
SecretTypeField = "type"

`PodHostField` has been widely used in `fields.OneTermEqualSelector(metav1.PodHostField, node.Name)`
`SecretTypeField`   will break the comment 

#https://github.com/kubernetes/kubernetes/blob/release-1.19/cmd/kubeadm/app/cmd/token.go#L373-L378
#https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder.go#L43-L44

Are you sure to change them all?

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/field_constants.go]: > `PodHostField` has been widely used in `fields.OneTermEqualSelector(metav1.PodHostField, node.Name)`

There's only a single use outside of test files. It's fine to change those from a constant to an inline field selector.

> `SecretTypeField` will break the comment
> 
> #https://github.com/kubernetes/kubernetes/blob/release-1.19/cmd/kubeadm/app/cmd/token.go#L373-L378
> #https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder.go#L43-L44
> 
> Are you sure to change them all?

Yes, inlining the field selector is fine.

**Comment by damemi** [comment]
Bumping this again, seems previous review comments have been addressed

**Comment by damemi** [comment]
Bumping again, is there any outstanding API review for moving these constants?

**Comment by liggitt** [comment]
/approve

lgtm once rebased

**Comment by damemi** [comment]
@xiaoanyunfei would you be able to rebase this? Sorry for the delay

**Comment by xiaoanyunfei** [comment]
@damemi @liggitt   
I'm tired of fixing the conflict again and again, would you please review this pr as soon as possible, thank you!

**Comment by damemi** [comment]
@xiaoanyunfei I apologize for the long delay in this PR. Because it involves so many areas, we needed to get feedback from all affected areas. I see now that it is approved, and am sorry again for the lengthy review time.
/lgtm

**Pull request merged**
